libs/parallelism/execution/include/hpx/execution/algorithms/ensure_started.hpp

```diff
@@ -0,0 +1,377 @@
+                        error_types<std::variant>, std::exception_ptr>>;
+                std::variant<std::monostate, done_type, error_type, value_type>
+                    v;
+
+                using continuation_type =
+                    hpx::util::unique_function_nonser<void()>;
+                boost::container::small_vector<continuation_type, 1>
+                    continuations;
+
+            public:
+                template <typename S_,
+                    typename = std::enable_if_t<
+                        !std::is_same<std::decay_t<S_>, shared_state>::value>>
+                shared_state(S_&& s)
+                  : os(hpx::execution::experimental::connect(
+                        std::forward<S_>(s), ensure_started_receiver{this}))
+                {
+                }
+
+                ~shared_state()
+                {
+                    if (!start_called)
+                    {
+                        std::terminate();
+                    }
+
+                    if (--reference_count == 0)
+                    {
+                        delete this;
+                    }
+                }
+
+            private:
+                template <typename R>
+                struct done_error_value_visitor
+                {
+                    std::decay_t<R> r;
+
+                    void operator()(done_type)
+                    {
+                        hpx::execution::experimental::set_done(std::move(r));
+                    }
+
+                    void operator()(error_type const& e)
+                    {
+                        std::visit(error_visitor<R>{std::forward<R>(r)}, e);
+                    }
+
+                    void operator()(value_type const& ts)
+                    {
+                        std::visit(value_visitor<R>{std::forward<R>(r)}, ts);
+                    }
+
+                    void operator()(std::monostate)
+                    {
+                        std::terminate();
+                    }
+                };
+
+                void set_predecessor_done()
+                {
+                    predecessor_done = true;
+
+                    std::unique_lock<hpx::util::spinlock> l{mtx};
+                    if (!continuations.empty())
+                    {
+                        l.unlock();
+                        for (auto const& continuation : continuations)
+                        {
+                            continuation();
+                        }
+
+                        continuations.clear();
+                    }
+                }
+
+            public:
+                template <typename R,
+                    typename =
+                        std::enable_if_t<std::is_rvalue_reference_v<R&&>>>
+                void add_continuation(R&& r)
+                {
+                    if (predecessor_done)
+                    {
+                        std::visit(
+                            done_error_value_visitor<R>{std::move(r)}, v);
```

Comment:
Why moving the forwarding reference?

Reply:
It's constrained with `is_rvalue_reference` because I really only want to move the receiver here. Is there a clearer idiom for not treating `R&&` as a universal reference, but really only as an rvalue reference?

Reply:
FWIW, a simple `template <typename T> void add_continuation(T&) = delete;` would do the trick as well in which case we could avoid the explicit SFINAE. 

Also, while `std::is_rvalue_reference_v<R&&>` is doing the right thing, I'd would have expected to see `!std::is_lvalue_reference_v<R>` instead. But that's just a personal preference.

Reply:
> FWIW, a simple `template <typename T> void add_continuation(T&) = delete;` would do the trick as well in which case we could avoid the explicit SFINAE.

That's a neat trick, I'll give it a shot!

> Also, while `std::is_rvalue_reference_v<R&&>` is doing the right thing, I would have expected to see `!std::is_lvalue_reference_v<R>` instead. But that's just a personal preference.

I prefer saying what I want it to be rather than what I don't want it to be. But with the previous suggestion I won't need either so hopefully we'll both be happy then!

Reply:
> FWIW, a simple `template <typename T> void add_continuation(T&) = delete;` would do the trick as well in which case we could avoid the explicit SFINAE.

This might also improve error messages, BTW.